source/sliding_puzzle.py

```diff
@@ -124,7 +139,88 @@ def _join(self, row: List[PuzzlePiece]) -> List[str]:
     def draw(self) -> str:
         """Draw the puzzle in its current state."""
         output = []
-        for row in self.pieces:
+        for row in self.rows:
             output.extend(self._join(row))
 
         return '\n'.join(output)
+
+    def shuffle(self) -> None:
+        """Shuffle the puzzle by randomizing the order of the pieces"""
+        for row in self.rows:
+            random.shuffle(row)
+        puzzle_heigth, puzzle_width = len(self.rows), len(self.rows[0])
+        # empty one of the rows
+        x = random.choice(range(puzzle_heigth))
+        y = random.choice(range(puzzle_width))
+        self.rows[x][y].clear()
+
+    def _get_empty_piece_position(self) -> PiecePosition:
+        x, y, index = [
+            (x, y, piece.index)
+            for x, row in enumerate(self.rows)
+            for y, piece in enumerate(row)
+            if piece._data == []
+        ][0]
+        return PiecePosition(x=x, y=y, index=index)
+
+    def move_down(
+        self,
+    ) -> None:
+        """Move the piece above the empty box down"""
```

Comment:
It's not really a box per-say..
```suggestion
        """Move the piece above the empty piece down."""
```